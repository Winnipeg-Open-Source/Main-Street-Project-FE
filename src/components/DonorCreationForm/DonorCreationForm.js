import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Checkbox, Label } from 'pcln-design-system';
import Form from 'components/Form';
import AddressForm from 'components/AddressForm';
import Input from 'components/Input';
import useForm from 'hooks/useForm';

const initialState = {
    name: null,
    address: null,
    city: 'Winnipeg',
    province: 'MB',
    postalCode: null,
    phoneNumber: null,
    email: null,
    isSubscribed: false,
};

const StyledLabel = styled(Label)`
    display: flex;
    align-items: center;
`;

function DonorCreationForm ({ className, onSaveClick, onCancelClick }) {
    const { state, onChange, onToggle } = useForm(initialState);
    const {
        name,
        address,
        city,
        province,
        postalCode,
        phoneNumber,
        email,
        isSubscribed,
    } = state;

    const handleSave = () => onSaveClick(state);

    return (
        <Form
            className={className}
            onSaveClick={handleSave}
            onCancelClick={onCancelClick}
        >
            <Input
                id='name'
                name='name'
                placeholder='Name'
                value={name}
                onChange={onChange}
            />
            <AddressForm
                address={address}
                city={city}
                province={province}
                postalCode={postalCode}
                onAddressChange={onChange}
                onCityChange={onChange}
                onProvinceChange={onChange}
                onPostalCodeChange={onChange}
            />
            <Input
                id='phone-number'
                name='phoneNumber'
                placeholder='Phone Number'
                value={phoneNumber}
                onChange={onChange}
            />
            <Input
                id='email'
                name='email'
                placeholder='Email'
                value={email}
                onChange={onChange}
            />
            <StyledLabel fontSize={0}>
                <Checkbox
                    id='newsletter'
                    name='isSubscribed'
                    chekced={isSubscribed}
                    onChange={onToggle}
                />
                Subscribe to Email Newsletter
            </StyledLabel>
        </Form>
    );
}

DonorCreationForm.displayName = 'DonorCreationForm';

DonorCreationForm.propTypes = {
    className: PropTypes.string,
    onSaveClick: PropTypes.func,
    onCancelClick: PropTypes.func,
};

DonorCreationForm.defaultProps = {
    className: '',
};

export default memo(DonorCreationForm);
