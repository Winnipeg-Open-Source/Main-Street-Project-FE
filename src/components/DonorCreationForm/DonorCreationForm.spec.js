import React from 'react';
import { fireEvent } from '@testing-library/react';
import renderWithTheme from 'tests/helpers/renderWithTheme';
import DonorCreationForm from './DonorCreationForm';

jest.mock('hooks/useForm', () => () => ({
    state: {
        name: 'Costco',
        city: 'Winnipeg',
    },
    onChange: jest.fn(),
    onToggle: jest.fn(),
}));

describe('DonorCreationForm', () => {
    it('renders without crashing', () => {
        const { asFragment } = renderWithTheme(<DonorCreationForm />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should save the current state', () => {
        const mockSave = jest.fn();
        const { getByText } = renderWithTheme(
            <DonorCreationForm onSaveClick={mockSave} />
        );

        const SaveButton = getByText('Save');
        expect(mockSave).toHaveBeenCalledTimes(0);
        fireEvent.click(SaveButton);
        expect(mockSave).toHaveBeenCalledTimes(1);
        expect(mockSave).toHaveBeenCalledWith({
            name: 'Costco',
            city: 'Winnipeg',
        });
    });
});
