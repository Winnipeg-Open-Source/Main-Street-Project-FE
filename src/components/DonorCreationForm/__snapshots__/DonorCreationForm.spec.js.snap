// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonorCreationForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-pbJYR hNzbLI sc-pkfPC fQMEfn"
  >
    <div
      class="sc-AxjAm sc-AxiKw sc-qQxFs dNNRWZ"
    >
      <div
        class="sc-AxjAm sc-AxiKw sc-qYUWV sa-don"
      >
        <input
          class="sc-pTWqp sc-pBlxj iLNgzO"
          font-size="2,,1"
          id="name"
          name="name"
          placeholder="Name"
          value="Costco"
        />
        <input
          class="sc-pTWqp sc-pBlxj iLNgzO"
          font-size="2,,1"
          id="address"
          name="address"
          placeholder="Street Address"
          value=""
        />
        <div
          class="sc-AxjAm sc-AxiKw ivVODM"
          width="1"
        >
          <input
            class="sc-pTWqp sc-pBlxj bRQLIM"
            font-size="2,,1"
            id="city"
            name="city"
            placeholder="City"
            value="Winnipeg"
          />
          <label
            class="sc-oVpqz cpWgdv className"
            color="border.light"
            font-size="10px"
            font-weight="bold"
          >
            <div
              class="sc-AxjAm sc-AxiKw iSVrPI"
              width="1"
            >
              <select
                class="sc-pZCuu kQVKTB sc-oTaid sc-pJipy ceCay"
                data-testid="province-select"
                font-size="2,,1"
                id="province"
                name="province"
              >
                <option
                  value="AB"
                >
                  Alberta
                </option>
                <option
                  value="BC"
                >
                  British Columbia
                </option>
                <option
                  value="MB"
                >
                  Manitoba
                </option>
                <option
                  value="NB"
                >
                  New Brunswick
                </option>
                <option
                  value="NL"
                >
                  Newfoundland and Labrador
                </option>
                <option
                  value="NS"
                >
                  Nova Scotia
                </option>
                <option
                  value="ON"
                >
                  Ontario
                </option>
                <option
                  value="PE"
                >
                  Prince Edward Island
                </option>
                <option
                  value="QC"
                >
                  Quebec
                </option>
                <option
                  value="SK"
                >
                  Saskatchewan
                </option>
                <option
                  value="NT"
                >
                  Northwest Territories
                </option>
                <option
                  value="NU"
                >
                  Nunavut
                </option>
                <option
                  value="YT"
                >
                  Yukon
                </option>
              </select>
              <svg
                aria-hidden="true"
                class="sc-Axmtr gXMRoh sc-fzqPZZ sc-pRFHb JnCiE"
                color="text.light"
                fill="currentcolor"
                focusable="false"
                height="24"
                role="img"
                tabindex="-1"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M7.4 8l4.6 4.6L16.6 8 18 9.4l-6 6-6-6L7.4 8z"
                />
              </svg>
            </div>
          </label>
        </div>
        <input
          class="sc-pTWqp sc-pBlxj iLNgzO"
          font-size="2,,1"
          id="postal-code"
          name="postalCode"
          placeholder="Postal Code"
          value=""
        />
        <input
          class="sc-pTWqp sc-pBlxj iLNgzO"
          font-size="2,,1"
          id="phone-number"
          name="phoneNumber"
          placeholder="Phone Number"
          value=""
        />
        <input
          class="sc-pTWqp sc-pBlxj iLNgzO"
          font-size="2,,1"
          id="email"
          name="email"
          placeholder="Email"
          value=""
        />
        <label
          class="sc-oVpqz sc-pRgDJ eInwpm"
          color="border.light"
          font-size="0"
          font-weight="bold"
        >
          <div
            class="sc-AxjAm sc-qapxP gFfLYW"
            color="primary"
          >
            <input
              class="sc-oTmHo gacPRn"
              color="primary"
              id="newsletter"
              name="isSubscribed"
              role="checkbox"
              size="20"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="sc-Axmtr cTdHPN sc-fzoXzr fDlmZK"
              data-name="checked"
              fill="currentcolor"
              focusable="false"
              height="24"
              role="img"
              tabindex="-1"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M6 3h12c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3H6c-1.7 0-3-1.3-3-3V6c0-1.7 1.3-3 3-3zm4 14l9-8.6L17.6 7 10 14.3l-3.6-3.5L5 12.2l5 4.8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="sc-Axmtr cTdHPN sc-fzpjYC kZSQFZ"
              data-name="empty"
              fill="currentcolor"
              focusable="false"
              height="24"
              role="img"
              tabindex="-1"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M6 5c-.6 0-1 .4-1 1v12c0 .6.4 1 1 1h12c.6 0 1-.4 1-1V6c0-.6-.4-1-1-1H6zm0-2h12c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3H6c-1.7 0-3-1.3-3-3V6c0-1.7 1.3-3 3-3z"
              />
            </svg>
          </div>
          Subscribe to Email Newsletter
        </label>
      </div>
      <div
        class="sc-AxjAm sc-AxiKw sc-psDhf iCfenm"
      >
        <button
          class="sc-oUOMp coOwBP"
          color="primary"
          width="1"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
